### PR TITLE
Fix incorrect string indentation in macro defs with `format_strings`

### LINKED ISSUE #4036: bizarre string indentation with `format_strings` and `wrap_comments`
On the latest rustfmt with a `.rustfmt.toml` of
```
format_strings = true
wrap_comments = true
```
and the code
```
macro_rules! test {
    () => {
        fn from() {
            None.expect(
                "We asserted that `buffer.len()` is exactly `$n` so we can expect \
                `ApInt::from_iter` to be successful.",
            )
        }
    };
}
```
This gets reformatted to 
```
macro_rules! test {
    () => {
        fn from() {
            None.expect(
                "We asserted that `buffer.len()` is exactly `$n` so we can expect \
         `ApInt::from_iter` to be successful.",
            )
        }
    };
}
```
Because there is a space in there and not a whole number of indentations, I think it is some off-by-one error in some code that deals with the indentations.

### PR DESCRIPTION
**Pull request by yodaldevoid**
Fixes #4036

### CODE REVIEW

File: src/utils.rs [hunk 1]
```diff
@@ -647,8 +647,9 @@ pub(crate) fn trim_left_preserve_layout(
 
 /// Based on the given line, determine if the next line can be indented or not.
 /// This allows to preserve the indentation of multi-line literals.
-pub(crate) fn indent_next_line(kind: FullCodeCharKind, _line: &str, config: &Config) -> bool {
-    !(kind.is_string() || (config.version() == Version::Two && kind.is_commented_string()))
+pub(crate) fn indent_next_line(kind: FullCodeCharKind, line: &str, config: &Config) -> bool {
+    !((kind.is_string() && !(config.format_strings() && line.ends_with('\\')))
+        || (config.version() == Version::Two && kind.is_commented_string()))
```
Comment (ytmimi) [inline src/utils.rs:652]: Thanks for submitting this PR!

Presumably by making this change ``indent_next_line`` returns true for the input referenced in the original issue (the input you included as a test case). Just trying to wrap my head around the added conditional check and I was hoping you could give a little more detail regarding the logic of your solution. Maybe a clarifying comment would be useful here?

Comment (yodaldevoid) [reply src/utils.rs:657]: I added a comment to the function explaining the added conditional. Do you think I should split the conditional up, say into an `if/else` statement to potentially make it clearer?

Comment (ytmimi) [reply src/utils.rs:657]: Thanks for taking the time to add some additional context to the doc comment. I'm still a little confused by the ``line.ends_with('\\')`` check and why that was needed. I should have been more descriptive about what I was confused by. Also, could you add an additional test case where the string is 2x or 3x longer. I want to see what happens when we need to wrap more than once.

Comment (yodaldevoid) [reply src/utils.rs]: I've refactored the function to be an `if/else` which should be a bit more readable when coupled with the added comments.

I've also added a test with lorem ipsum as the string which was long enough to require wrapping a few times.

Comment (ytmimi) [reply src/utils.rs]: Thanks a lot for the quick turn around on this. I think the current implementation (using ``if/else``), along with the added comment gives plenty of context for what's going on here.

@calebcartwright I'm happy to move forward with these changes, but please let us know if there's anything else you'd like to see from this PR.

File: CHANGELOG.md [hunk 2]
```diff
@@ -40,6 +40,7 @@
 - Handles cases where certain control flow type expressions have comments between patterns/keywords and the pattern ident contains the keyword [#5009](https://github.com/rust-lang/rustfmt/issues/5009)
 - Handles tuple structs that have explicit visibilities and start with a block style comment [#5011](https://github.com/rust-lang/rustfmt/issues/5011)
 - Handles leading line-style comments in certain types of macro calls [#4615](https://github.com/rust-lang/rustfmt/issues/4615)
+- bizarre string indentation in macro defs with `format_strings` [#4036](https://github.com/rust-lang/rustfmt/issues/4036)
```
Comment (camelid) [inline CHANGELOG.md:43]: Did you mean to write "Fixes bizarre string ..."?

Comment (yodaldevoid) [reply CHANGELOG.md]: Good catch

File: CHANGELOG.md [hunk 3]
```diff
@@ -40,6 +40,7 @@
 - Handles cases where certain control flow type expressions have comments between patterns/keywords and the pattern ident contains the keyword [#5009](https://github.com/rust-lang/rustfmt/issues/5009)
 - Handles tuple structs that have explicit visibilities and start with a block style comment [#5011](https://github.com/rust-lang/rustfmt/issues/5011)
 - Handles leading line-style comments in certain types of macro calls [#4615](https://github.com/rust-lang/rustfmt/issues/4615)
+- Fixes bizarre string indentation in macro defs with `format_strings` [#4036](https://github.com/rust-lang/rustfmt/issues/4036)
```
Comment (calebcartwright) [inline CHANGELOG.md:43]: I appreciate you thinking to include an entry in the changelog! However, we'll need to remove, or move+change this.

The section you've added this to is actually for the already-released 1.4.38 version, so we shouldn't add notes for this fix to a version where it doesn't exist. Secondly, we don't want to use vague, subjective language and adjectives like those typically provided by frustrated users in issue reports. Users reading this type of line in a changelog can't really glean much information from it, which is instead why try to frame the wording in our changelog in more of a functional manner that describes the behavior that was fixed in a way anyone could understand (e.g. `Fixed issue in macro defs where every line but the first line in a multiline string was missing an indent level when ``format_strings`` was enabled`).

If you'd like to keep an entry (which would indeed be helpful for me whenever it comes time for the next release), let's adjust the wording a bit and move to the top under the `Unreleased` section of the page

File: tests/target/issue-4036.rs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+// rustfmt-format_strings: true
+
```
Comment (calebcartwright) [inline tests/target/issue-4036.rs:2]: Realize this wasn't in the report, but could you do me a favor and add another set of tests that have `hard_tabs` turned on? We've got some idempotency issues within macro defs in some similar looking cases, and I'd like to rule out those occurring here (or if they do, at least confirm it's nothing new/same behavior could be produced on master)

**Review by calebcartwright — COMMENTED** [review]
LGTM, thanks! Couple minor asks left inline below, otherwise good to go

File: CHANGELOG.md (hunk 3 above)
Comment (yodaldevoid) [reply CHANGELOG.md:43]: No problem. I thought I had placed it under `Unreleased`, but I obviously wasn't paying attention.

File: tests/target/issue-4036.rs (hunk 4 above)
Comment (yodaldevoid) [reply tests/target/issue-4036.rs]: I've added a test with `hard_tabs` enabled. Let me know if you would like the test modified or additional tests added.

**Comment by calebcartwright** [comment]
Awesome, thank you!

**Pull request merged**